Cohort course-wide and inline discussion topics[UI].

muzaffaryousaf opened the pull request:
Ticket # https://openedx.atlassian.net/browse/TNL-1256 

This adds the UI to the cohort management tab on LMS side. It allows the staff to cohort the course-wide as well as, inline discussions via friendly UI. 

The Sandbox is available at: http://cohort-discussions.m.sandbox.edx.org/

catong commented:
Hi @muzaffaryousaf It looks like the sandbox does not include the change to the Cohorts tab? It still has the label "Cohort Management" and it in the right-most position (rather than being ext to Membership). Can you please provide a final sandbox if this is not the most up to date? Thanks!

muzaffaryousaf commented:
@catong i've updated the sand-box. Thanks for heads up.

muzaffaryousaf commented:
Just FYI, As we were implementing this feature parallel to the other cohort related functionality, so there are 3 PR's for this feature. This https://github.com/edx/edx-platform/pull/7321 is latest  and final. Also,  this feature is reviewed by @symbolist @catong @marcotuts and @explorerleslie .

For review comments you can look to these Closed PRs: 
1- https://github.com/edx/edx-platform/pull/7108
2 - https://github.com/edx/edx-platform/pull/7199

Thanks.

muzaffaryousaf commented:
test this please

symbolist commented on common/lib/django_test_client_utils.py:
```diff
@@ -0,0 +1,46 @@
+
+
+setattr(RequestFactory, 'patch', request_factory_patch)
+setattr(Client, 'patch', client_patch)
```
I think you should check if the attribute is already set. So when we upgrade to the next version of django, the django version would start getting used automatically.

symbolist commented on lms/static/js/groups/models/cohort_discussions.js:
```diff
@@ -0,0 +1,10 @@
+    edx.groups = edx.groups || {};
+
+    edx.groups.DiscussionTopicsModel = Backbone.Model.extend({
+    });
```
Can you add the parameters that will be set on this model? Helps in understanding.

symbolist commented on lms/static/js/factories/cohorts_factory.js:
```diff
@@ -1,33 +1,40 @@
 ;(function (define, undefined) {
```
I think a better path for this file will be groups/views/cohorts_dashboard_factory.js.

symbolist commented on lms/djangoapps/django_comment_client/utils.py:
```diff
@@ -134,9 +134,7 @@ def _sort_map_entries(category_map, sort_alpha):
-def get_discussion_category_map(course):
-    course_id = course.id
-
+def get_discussion_category_map(course, ignore_always_cohort_inline_discussions=False):
```
Can you add a docstring explaining the purpose of the parameters?

Also I am thinking, when fetching the map for the instructor dashboard besides ignoring the `always_cohort_inline_dicussions` setting, shouldn't the `is_cohorted` setting also be ignored? Or in other words a topic should only be marked `is_cohorted` if it is in the list.

If yes, you can call the parameter something like `cohorted_if_in_list`.

symbolist commented:
Besides a few nitpicks :+1: from me!

muzaffaryousaf commented:
@andy-armstrong can you please review the code for this PR ?
Thanks.

andy-armstrong commented:
@explorerleslie What happened to the idea of a confirmation before disabling cohorts in a live course? Did we decide to hold off on that? I just verified that I can still disable it with one click on the sandbox.

explorerleslie commented:
@andy-armstrong it is not a part of this PR, but in a separate story: https://openedx.atlassian.net/browse/TNL-1589

andy-armstrong commented on common/test/acceptance/pages/lms/instructor_dashboard.py:
```diff
@@ -452,6 +456,109 @@ def is_cohorted(self, state):
+        """
+        save_button_css = '%s %s' % (self.discussion_form_selectors[key], '.action-save')
+        disabled = self.q(css=self._bounded_selector(save_button_css)).attrs('disabled')
+        if disabled[0] == 'true':
```
This would be more concise and arguably easier to read as:

    return disabled[0] == 'true'

andy-armstrong commented on common/test/acceptance/pages/lms/instructor_dashboard.py:
```diff
@@ -452,6 +456,109 @@ def is_cohorted(self, state):
 
+    def toggle_discussion_topics(self):
+        """
+        Shows the discussion topics.
```
I'd change this to say "Toggles showing of discussion topics", since it doesn't always show them.

andy-armstrong commented:
Why does the Cohorts tab still show "Edit settings in Studio"?

![image](https://cloud.githubusercontent.com/assets/5985072/6676420/17078cd6-cbfe-11e4-9a1b-c640abee3e20.png)

explorerleslie commented:
whoops, we missed that, thanks @andy-armstrong. @muzaffaryousaf, can you please remove the "Edit settings in Studio" link as part of this PR?

andy-armstrong commented:
It took me a while to find how to set the discussions. The "Specify whether discussion topics are divided by cohort" link doesn't seem prominent enough for such an important part of the UI. When you have a cohort selected it really just slides off the page and looks like a help link. I know we had a similar issue with the CSV upload, but this one seems worse to me somehow.

The link also sounds like it is a single setting (specifying whether topics are cohorted), rather than expanding a section that gives you a lot of controls for managing cohorts. Could it be "Specify which discussion topics are cohorted" maybe? @catong what do you think?

andy-armstrong commented:
The content-specific discussion topics section is confusing when you don't have any:

![image](https://cloud.githubusercontent.com/assets/5985072/6676561/4a0a5748-cbff-11e4-9901-92719267e1fa.png)

Maybe this radio button should be disabled in this situation.

explorerleslie commented:
@marcotuts and @catong, for @andy-armstrong 's comment above, what would you think about just putting "No content specific discussion topics in course" or something? Then at least we let the course author specify the always cohort inline discussions setting to false before they add any topics.

andy-armstrong commented:
The list of discussion topics needs to scroll, because it makes it very hard to find the "Save" button if you have more than ten or so topics. I believe that there are courses with many more than this.

![image](https://cloud.githubusercontent.com/assets/5985072/6676867/87573452-cc01-11e4-875e-6e16a60da1c8.png)

andy-armstrong commented:
I would recommend hiding the list of topics when "Always cohort..." is on. Otherwise there is a long disabled tree that the user has to scroll past to see if there are any settings further down.

andy-armstrong commented:
The list of course-wide topics probably should also be scrolled if it gets too long.

catong commented:
@explorerleslie (CC @marco @andy-armstrong) I like the idea of indicating clearly that there are no content-specific discussion topics. 
Thinking through the scenario of someone deliberately setting "Cohort some..." before they go to create topics. Hmm, this might be confusing. We still need a default setting that a content-specific discussion topic is set to when it's first created, and that should be the state that users see when they come to this page to manage discussions after creating them. I think having "Always cohort..." set by default when no content-specific discussion topics exist is a clearer behavior than allowing them to select either option BEFORE creating any topics. With the current behavior (and the addition of some note indicating that there are currently no content-specific discussion topics) it's clear to users what the default setting is. After creating topic they can then come here, see and understand the default setting, and make a single click to change the option and deselect the discussion topics they don't want cohorted. My rather convoluted 2 cents...

andy-armstrong commented:
I keep clicking on the labels for the radio and check buttons and find it frustrating that they don't select. They actual radio and check buttons themselves are very small and hard to hit, especially when you are trying to move up and down and select multiples at a time.

andy-armstrong commented:
I wonder if we need a "Select All" option (and corresponding "Deselect All") to make it easier to manage a large list of course-wide topics. We might also want the ability to set the default setting for new topics. It is frustrating to have to go back and forth between Studio and the Instructor Dashboard to add a new cohorted topic.

andy-armstrong commented:
@muzaffaryousaf I'm done playing with the sandbox and the feature is working great. I'm out of time for tonight so I'll review the code tomorrow.

muzaffaryousaf commented:
test this please

explorerleslie commented:
@marcotuts -- what do you think about making the list of discussion topics scroll? I'm tempted to not change right now. I don't think it's hugely disruptive to not have in-section scrolling -- the page does scroll and the save button is at the bottom, and I'm worried about making the instructor dashboard more complicated than it already is, but I wanted to make sure you had a chance to respond to @andy-armstrong 's feedback.

explorerleslie commented:
@andy-armstrong we discussed at length whether we should hide the list of topics when "Always cohort" was selected, but ultimately decided against it, because it was helpful to have the context of the discussion topics in your view when deciding between "Always cohort" and "Select...". We can always hide it later if we get user feedback that it's confusing.

explorerleslie commented:
@muzaffaryousaf can you please make the labels for checkboxes and radio buttons clickable, in addition to the actual checkbox/radio button?

explorerleslie commented:
@andy-armstrong we decided against a select-all/deselect-all option during design. The "Always cohort" and "select cohort..." options set the defaults for new topics as well as give you a way to always cohort or only select a few for non-cohorting. Predominant use case for course teams is still for cohorting all inline discussions, because oftentimes inline discussions get a ton of traffic, in which case it makes sense to cohort most/all inline discussions.

explorerleslie commented:
ah, @muzaffaryousaf, it looks like you already added a scrollbar for scrolling topics. Can you please remove that? 

In summary, after discussing with @marcotuts, the changes you should implement based on @andy-armstrong 's feedback are:
- remove the "Edit settings in Studio" link as part of this PR
- make the labels for checkboxes and radio buttons clickable, in addition to the actual checkbox/radio button
- Add text that says "No content-specific discussion topics." when none are present in course. 

Do NOT change the following items:
- Don't make the list of topics scroll
- Don't implement select-all/deselect-all
- Don't hide the list of topics when "Always cohort..." is selected

catong commented:
@explorerleslie Can I suggest that the wording for "No content-specific discussion topics" should parallel the wording for what we say when there are no cohorts or no content groups. 
I think we should say "No content-specific discussion topics exist", but @muzaffaryousaf if you can see the other labels in the code, please make it whatever is parallel?

muzaffaryousaf commented:
@explorerleslie alright, i will remove the scroll bars. Thanks for details.

explorerleslie commented:
@catong of course, I've edited my comment above. @muzaffaryousaf FYI.

andy-armstrong commented:
@muzaffaryousaf I'm seeing a 500 error in discussions and the cohorts tab after uploading the MIT 6.002x course. I don't know if this is a bug in the uploader on an issue with this PR so I figured I'd let you know. It appears to be an issue with this service:

http://cohort-discussions.m.sandbox.edx.org/courses/MITx/6.002x/1T2015/cohorts/settings

andy-armstrong commented:
@explorerleslie @marcotuts Regarding scrolling: I tried to upload 6.002x to see how the Cohorts tab would cope, but it breaks the server currently. However, I was able to get Justice to load which has a fairly large number:

http://cohort-discussions.m.sandbox.edx.org/courses/HarvardX/ER22.1x/1T2014/instructor#view-cohort_management

I wonder whether it would be better to make the topic sections initially collapsed so then you would only get an item per section. This would fit with the notion that most of the time cohorting is enabled or disabled for an entire section. Of course this would require that this be implemented as more of a tree that can expand and collapse which is more work.

andy-armstrong commented on common/test/acceptance/tests/discussion/test_cohort_management.py:
```diff
@@ -640,6 +624,201 @@ def _verify_cohort_by_csv_notification(self, expected_message):
+        When I click on the course-wide discussion topic
+        Then I see the enabled save button
+        When I click on save button
+        Then I see success message.
```
I'd add something like:

    When I reload the page
   Then I see the discussion topic selected

andy-armstrong commented on common/test/acceptance/tests/discussion/test_cohort_management.py:
```diff
@@ -640,6 +624,201 @@ def _verify_cohort_by_csv_notification(self, expected_message):
+        And a inline discussion topic with disabled Save button.
+        When I click on always_cohort_inline_topics
+        Then I see enabled save button
+        And I see disabled inline discussion topics.
```
Again I'd include reloading the page and verifying that the save worked. In fact, that goes for all of these tests.

andy-armstrong commented on common/test/acceptance/tests/discussion/test_cohort_management.py:
```diff
@@ -640,6 +624,201 @@ def _verify_cohort_by_csv_notification(self, expected_message):
+        cohorted_topics_after = self.cohort_management_page.get_cohorted_topics_count(self.inline_key)
+        self.assertNotEqual(cohorted_topics_before, cohorted_topics_after)
+
+    def test_toggle_inline_discussion_topics(self):
```
It would be good to add some more tests around the partial selection of child topics:
* verify that selecting a subset of children causes the category to be marked as partially selected
* verify that deselecting the final child deselects the category
* verify that selecting all the children fully selects the category

andy-armstrong commented on common/test/acceptance/tests/discussion/test_cohort_management.py:
```diff
@@ -640,6 +624,201 @@ def _verify_cohort_by_csv_notification(self, expected_message):
 
 
 @attr('shard_3')
+class CohortDiscussionTopicsTest(UniqueCourseTest, CohortTestMixin):
```
These are excellent BokChoy tests!

andy-armstrong commented on lms/djangoapps/django_comment_client/utils.py:
```diff
@@ -184,12 +222,21 @@ def get_discussion_category_map(course):
                 node[level]["start_date"] = category_start_date
 
+        # pylint: disable=invalid-name
+        always_cohort_inline_discussions = (not cohorted_if_in_list and
```
I'd break this line after the first ``(``.

andy-armstrong commented on lms/static/js/groups/views/cohort_discussions.js:
```diff
@@ -0,0 +1,114 @@
+         Args:
+            $element (JQuery element): The element to disable/enable.
+            disable (Bool): The flag to add/remove 'disabled' attribute.
+        **/
```
Minor nit: this isn't a valid JSDoc comment. It needs to look like these:

http://usejsdoc.org/about-getting-started.html

Each line starts with ``* `` and then the final line ends with ``*/``.

We aren't yet generating documentation from our JavaScript, but I hope we do relatively soon.

andy-armstrong commented on lms/static/js/groups/views/cohort_discussions.js:
```diff
@@ -0,0 +1,114 @@
+            disable (Bool): The flag to add/remove 'disabled' attribute.
+        **/
+        setDisabled: function($element, disable) {
+            if (disable) {
```
This would be easier to read as:

    $element.prop('disabled', disable ? 'disabled' : false);

catong commented:
@explorerleslie @muzaffaryousaf  @marcotuts , @andy-armstrong 's review comments have brought up some great points. As a result of a chat this morning with Leslie I realized that I hadn't fully realized (and thus hadn't documented) that selecting "Cohort some content-specific discussion topics" changes the default setting, so that any content-specific discussion topics are created as unified discussion topics. I am going to make updates to the doc to remedy this. We also talked about reworking the labels (sorry, @muzaffaryousaf !) to make this effect of changing the setting more obvious to users

@explorerleslie I think your suggestion is the best so far: 
replace "Cohort some content-specific discussion topics" 
with "Cohort only selected content-specific discussion topics"

Thoughts from anyone else?

andy-armstrong commented on lms/static/js/groups/views/cohort_discussions.js:
```diff
@@ -0,0 +1,114 @@
+
+    edx.groups = edx.groups || {};
+
+    edx.groups.CohortDiscussionsView = Backbone.View.extend({
```
I think the name ``CohortDiscussionsView`` is confusing as this isn't a view that shows discussions. Maybe it should be ``CohortDiscussionConfigurationView`` or something similar.

andy-armstrong commented on lms/static/js/groups/models/cohort_discussions.js:
```diff
@@ -0,0 +1,14 @@
+
+    edx.groups = edx.groups || {};
+
+    edx.groups.DiscussionTopicsModel = Backbone.Model.extend({
```
This isn't really a model of discussion topics. It is a model of which discussion topics are cohorted. Maybe ``DiscussionTopicsSettingsModel``?

andy-armstrong commented on lms/static/js/groups/views/cohort_discussions.js:
```diff
@@ -0,0 +1,114 @@
+            cohortSettingsModel.save(
+                fieldData, {patch: true, wait: true}
+            ).done(function () {
+                    saveOperation.resolve();
```
This indentation is a little off. It should only be four spaces indented from the line before it.

andy-armstrong commented on lms/static/js/groups/views/cohort_discussions_course_wide.js:
```diff
@@ -0,0 +1,84 @@
+
+    edx.groups = edx.groups || {};
+
+    edx.groups.CourseWideDiscussionsView = CohortDiscussionsView.extend({
```
This should also be named ...DiscussionsConfigurationView.

andy-armstrong commented on lms/static/js/groups/views/cohort_discussions_inline.js:
```diff
@@ -0,0 +1,148 @@
+        },
+
+        /**
+         Disables the discussion category checkboxes.
```
I found this comment confusing. It would be better if it had a single sentence saying what it is used for, and then have bullets for all of the actions that it does.

andy-armstrong commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -54,23 +67,71 @@ define(['backbone', 'jquery', 'js/common_helpers/ajax_helpers', 'js/common_helpe
+
+            createMockCohortDiscussionsJson = function () {
+                return {
+                            course_wide_discussions: {
```
This should be indented four spaces across from the ``return``, not from the ``{``

andy-armstrong commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -960,5 +1072,206 @@ define(['backbone', 'jquery', 'js/common_helpers/ajax_helpers', 'js/common_helpe
+
+                describe("Course Wide", function() {
+
+                    it("save button is disabled initially", function() {
```
The idea of Jasmine test names is that they read as sentences starting with the ``it`` at the beginning. This could be something like:

    it('initially shows the "Save" button as disabled', ...

Also, you are mixing single quotes with double quotes for your strings. Our JS convention is to always use single quotes.

muzaffaryousaf replied:
Thanks for pointing out the name for tests.

About the single vs double quotes, there are many or you can say most of the existing specs are using double quotes instead of single. i was confused to use either single or double, so i prefer to use double because they are being used in majority of the places. Also, i've checked from jasmine doc, they are also using double quotes for specs. http://jasmine.github.io/2.0/introduction.html

muzaffaryousaf replied:
@andy-armstrong i will use the single quotes instead if it's our convention. Is there any doc for such conventions ? please share if any. It would be helpful for future work.

andy-armstrong replied:
@muzaffaryousaf Good question. Here's the JS guidelines but it is out-of-date and doesn't describe single quotes. In general, following the style of the file you are in makes sense, so if all the other tests are using double quotes then it is reasonable to do the same. 

https://github.com/edx/edx-platform/wiki/Javascript-standards-for-the-edx-platform

I've added links to all our coding guidelines on the T&L code review guidelines wiki:

https://openedx.atlassian.net/wiki/display/TNL/Code+Review+Guidelines+and+Gotchas

andy-armstrong commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -960,5 +1072,206 @@ define(['backbone', 'jquery', 'js/common_helpers/ajax_helpers', 'js/common_helpe
+                        $(courseWideView.$('.check-discussion-subcategory-course-wide')[1]).prop('checked', 'checked').change();
+                        courseWideView.$('.action-save').click();
+
+                        AjaxHelpers.respondWithError(requests, 500);
```
I'd add a test for what happens when the PATCH succeeds but the subsequent GET returns a 500.

andy-armstrong commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -960,5 +1072,206 @@ define(['backbone', 'jquery', 'js/common_helpers/ajax_helpers', 'js/common_helpe
+                        enableSaveButton();
+                    });
+
+                    it("save the topics", function() {
```
I'd add tests for selecting and deselecting child topics and verifying that the correct three-way behavior is demonstrated on the category.

andy-armstrong commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -960,5 +1072,206 @@ define(['backbone', 'jquery', 'js/common_helpers/ajax_helpers', 'js/common_helpe
+                        expect(inlineView.$(inlineDiscussionsSaveButtonCss).prop('disabled')).toBeTruthy();
+                    });
+
+                    it("always cohort radio button is selected", function() {
```
Add tests for switching back and forth between always cohort and manually cohort.

andy-armstrong commented on lms/templates/instructor/instructor_dashboard_2/cohort-discussions-inline.underscore:
```diff
@@ -0,0 +1,30 @@
+                    <p><%- gettext('Specify whether content-specific discussion topics are divided by cohort.') %></p>
+                    <div class="always_cohort_inline_discussions">
+                        <input type="radio" name="inline" class="check-all-inline-discussions" <%- alwaysCohortInlineDiscussions ? 'checked="checked"' : '' %>/>
+                        <span class="all-inline-discussions"><%= interpolate('Always cohort content-specific discussion topics') %></span>
```
Why are you using ``interpolate`` here? I think this should be ``gettext``.

andy-armstrong replied:
You should also use ``<%-`` so that translations with Unicode characters will be escaped correctly.

andy-armstrong commented on lms/templates/instructor/instructor_dashboard_2/cohort-discussions-inline.underscore:
```diff
@@ -0,0 +1,30 @@
+                    </div>
+                    <div class="cohort_inline_discussions">
+                        <input type="radio" name="inline" class="check-cohort-inline-discussions" <%- alwaysCohortInlineDiscussions ? '' : 'checked="checked"' %>/>
+                        <span class="all-inline-discussions"><%= interpolate(' Cohort some content-specific discussion topics') %></span>
```
This should also be ``gettext``.

andy-armstrong replied:
Also use ``<%-``

andy-armstrong commented on lms/templates/instructor/instructor_dashboard_2/cohort-group-header.underscore:
```diff
@@ -18,9 +18,4 @@
-            <a href="<%= studioAdvancedSettingsUrl %>" class="action-secondary action-edit"><%- gettext("Edit settings in Studio") %></a>
-        <% } %>
-    </div>
 </div>
```
Missing newline

andy-armstrong commented:
@muzaffaryousaf I'm finished with my first review. This is looking fantastic. Please tag me when this is ready for a second review, and let me know if you have any questions.

catong commented:
@muzaffaryousaf, as a result of discussion with @explorerleslie today, we realized we should make a tweak to the name of the "Cohort some content-specific discussion topics" label. Can you please change it to "Cohort selected content-specific discussion topics"? Thanks!

muzaffaryousaf commented:
test this please

muzaffaryousaf commented:
test this please

muzaffaryousaf commented:
@catong @explorerleslie i've addressed all the comments and updated the code on sandbox, please have a look and let me know, if i'm missing something. 
Thanks

muzaffaryousaf commented:
@andy-armstrong i've addressed all of your concerns, can you please re-review the changes?
Thanks

catong commented:
Thanks, @muzaffaryousaf for the quick updates!
The revised "Cohort selected content-specific discussion topics" label looks good to me.
@explorerleslie @marcotuts can you please comment on the implementation of the message when no content-specific discussion topics exist? I think that when we chatted yesterday, @explorerleslie, I/we envisioned still seeing the section and the controls for content-specific discussion topics, and simply having the "No content-specific discussion topics exist" appear in place of the list of topics, but perhaps that is not possible/not in keeping with our patterns?

andy-armstrong commented:
@catong I agree with you. The current message is a little out of context by itself. If I don't know that there's meant to be a content-specific discussion topic control there, then it is very confusing. I'd rather see it as a message in place of the list of topics, but then disable the radio button so that it isn't possible to select manual control.

![image](https://cloud.githubusercontent.com/assets/5985072/6716245/62445300-cd7a-11e4-8c29-5e443713e16b.png)

andy-armstrong commented:
@muzaffaryousaf @explorerleslie an interesting problem that I just hit is that you can't see the list of content-specific discussion topics if the course hasn't gone live yet. I had to switch the release date to be in the past to be able to specific this cohorting. That seems like it will be a problem when an instructor is trying to set the course up ahead of its launch date.

My guess is that this is because it is using the same API that the discussion tab uses, and the discussion tab also doesn't let the user see topics that haven't been released yet. I think the instructor dashboard should see all topics regardless of release date.

andy-armstrong commented on lms/static/js/groups/views/cohort_discussions_course_wide.js:
```diff
@@ -0,0 +1,83 @@
+                self.model.fetch()
+                    .done(function () {
+                        self.render();
+                        self.showMessage(gettext('Changes Saved.'), self.$('.course-wide-discussion-topics'));
```
We typically don't use title case for our messages, or end with a period. Some examples: "Saved cohort", "The cohort can not be saved" etc. @catong would you agree?

catong replied:
@andy-armstrong yes, I agree we shouldn't use title case. I suggest: "Your changes have been saved" (no period)
I haven't been able to get a sense of whether using or not using periods is more prevalent in our UI when we have single sentences in our UI confirmation/error messages. If you know that it's more typical that we don't use periods then that's fine with me. (I just have to remember for the next recommendations I make! :-))

muzaffaryousaf replied:
@andy-armstrong i'm not sure about using the **.** . There are many single sentences ending with **.** that we are using on instructor dashboard. Some examples: "Please enter a student email address or username." , "Please enter a problem location." and many more.

andy-armstrong commented:
I can't seem to set one child topic to be cohorted and not the other. When I hit "Save" then they both end up being cohorted.

Before save:
![image](https://cloud.githubusercontent.com/assets/5985072/6716457/13067a6e-cd7c-11e4-8234-ea84e01f581f.png)

After save:
![image](https://cloud.githubusercontent.com/assets/5985072/6716465/1f4b53b2-cd7c-11e4-9286-852962124074.png)

You can see this in my test course: http://cohort-discussions.m.sandbox.edx.org/courses/AndyA/AA101/1/instructor#view-cohort_management

andy-armstrong commented:
@muzaffaryousaf Thanks for addressing my feedback. The changes all look great. The only thing is that I think you missed my suggestion for adding both Jasmine and Bok Choy tests for handling subsets of topics within a category. This doesn't seem to work for me on the server, so having tests in this area would be great. The other big issue is the inability to select content-specific discussion topics for a discussion that hasn't been released yet.

explorerleslie commented:
@muzaffaryousaf a few things: 

- Please make discussion topics on the instructor dashboard show at all times, even if the course start date is in the future or the inline discussion component hasn't been released yet (good catch @andy-armstrong)
- I think there was a miscommunication with the implementation of the "No content-specific discussion topics...." message. Please implement it so that the message we specified is between the two horizontal rule lines in this screenshot that was posted previously in the thread: https://cloud.githubusercontent.com/assets/5985072/6676561/4a0a5748-cbff-11e4-9901-92719267e1fa.png. Don't make it into its own warning message and hide the whole section. Let me know if you want to discuss (cc @marcotuts)
- Interestingly, I'm not seeing the select-one-select-all behavior in the demo course that @andy-armstrong saw in his test course, so definitely check out his course to see what's going on.

andy-armstrong commented:
@muzaffaryousaf I forgot to check about the 500 error on the Justice course, so I uploaded it again and it still has the same problem:

http://cohort-discussions.m.sandbox.edx.org/courses/MITx/600.2x/2014_T1/discussion/forum

If you go to the Cohorts tab then you'll see that the cohort section isn't shown and that there's a 500 response from this AJAX request:

http://cohort-discussions.m.sandbox.edx.org/courses/MITx/600.2x/2014_T1/cohorts/settings

Can you investigate to see whether this is a problem introduced by your changes please. It is possible that old courses aren't migrating correctly.

muzaffaryousaf commented:
@andy-armstrong Thanks for pointing this out.
I've investigated the 500 issue, seems like nothing is wrong with new/migration functionality. As per my investigation, it turned out that `get_discussion_category_map` is the main culprit  again. Here is the log 

```
modules = _get_discussion_modules(course)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/django_comment_client/utils.py", line 74, in _get_discussion_modules
    return filter(has_required_keys, all_modules)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/django_comment_client/utils.py", line 69, in has_required_keys
    if getattr(module, key) is None:
AttributeError: 'ErrorDescriptorWithMixins' object has no attribute 'discussion_id'
```

Mainly this line `if getattr(module, key) is None:` 
It tries to use `getattr` for `discussion_id` without giving the default value, resulting into *AttributeError*.

@explorerleslie i can fix this in same PR, although it's not related to this. Also, do we need to make a ticket for this.Please let me know what do you think.

muzaffaryousaf commented:
@explorerleslie do you want me to disable radio buttons in case of no content-specific topics ?

muzaffaryousaf commented:
@andy-armstrong @explorerleslie The issue about selecting/deselecting the subset of inline topics on andy's sandbox is because of duplicate IDs. I'm not sure how he end up with duplicate IDs.
Here is a attached screenshot. Please have a look.
![subcategory-topics](https://cloud.githubusercontent.com/assets/2994509/6729563/6bdf4d58-ce55-11e4-8a54-25daa8744dfb.png)

muzaffaryousaf commented:
@andy-armstrong i've already added the suggested tests except the reload behaviour, Now, i've added that just for the safe side.

muzaffaryousaf commented:
@explorerleslie here is a screenshot for no content-specific discussion topics. please let me know what do you think.

![no content](https://cloud.githubusercontent.com/assets/2994509/6730217/30e1fe7e-ce5c-11e4-8bf2-cc5cd8f062cc.png)

explorerleslie commented:
@muzaffaryousaf 

- The no content-specific discussion topics screenshot looks good. @marcotuts should also have a look. 
- The 500 error -- was it introduced by anything in the cohorts work you all have done this quarter? Either way, please create a new ticket in JIRA to track. If it's something introduced with cohorts work, we should fix before this goes live. If not, we can treat as lower priority.

andy-armstrong commented:
@muzaffaryousaf Thanks for tracking down the issue with selecting a subset of topics. I used the duplicate button in Studio to create all these discussions, so I guess there is a bug where the duplicate ends up with the same id as the original. @explorerleslie is this an issue that you are aware of?

muzaffaryousaf commented:
@andy-armstrong @explorerleslie i've addressed all of your comments and update the code on sandbox. Please have a look and let me know.
Thanks

andy-armstrong commented:
@muzaffaryousaf Looks great to me. Thanks for tackling all of my feedback. :+1: 

One minor comment: I don't see any tests around the new logic to include unreleased discussion topics. Maybe I missed them (there are a lot of tests which is fantastic!) but if not would you be able to add one. I manually verified that the behavior works for the case I reported yesterday, but it would be good to have a test for it too. No need to re-review just for that.

One other minor thing: I think the area beneath the radio buttons (where the topic list and/or message are shown) would be better if it was visually indented. I think that would make it clearer that the topic list relates to the radio button. @marcotuts does that sound good to you?

muzaffaryousaf commented:
@andy-armstrong thanks for such a thorough review and testing. About the test which you mentioned, I've added this already to check if the API returning discussion topics that are not started yet, you can see the test here https://github.com/edx/edx-platform/pull/7321/files#diff-eeac770f5a8e2e3d0c84e8cfee8a55fdR321 

Thanks

muzaffaryousaf commented:
@explorerleslie @catong do you have anything to add in this PR or we are good to merge it?

explorerleslie commented:
hi @andy-armstrong I was not aware that duplicating a discussion component duplicated the IDs. I'll file a bug. 

WRT the radio buttons, I suggested something similar yesterday to @marcotuts but he made the good case that tying them in that way introduces more issues -- at this point, I think it's better to launch with what we have and make changes as we see where users are confused because they'll definitely tell us if they are :)

explorerleslie commented:
@muzaffaryousaf :+1: from me!

catong commented:
@muzaffaryousaf :+1: from me as well. Thanks!

andy-armstrong commented:
@muzaffaryousaf I forgot to say: don't forget to squash your commits before merging.
